File: Languages/cpp/algorithms-dsa/leetcode/greedy/task_scheduler.cpp
/**
 * LeetCode 621: Task Scheduler
 * Topic: Greedy
 * Difficulty: Medium
 *
 * Given tasks and cooldown n, find minimum intervals to finish all tasks.
 * Count frequencies; the most frequent task dictates idle slots.
 * Time: O(t) where t = number of tasks, Space: O(1) (26 letters max)
 */
#include <iostream>
#include <vector>
#include <cassert>
#include <algorithm>
using namespace std;

int leastInterval(vector<char>& tasks, int n) {
    vector<int> freq(26, 0);
    for (char c : tasks) freq[c - 'A']++;
    int maxFreq = *max_element(freq.begin(), freq.end());
    int maxCount = count(freq.begin(), freq.end(), maxFreq);
    int partitions = maxFreq - 1;
    int emptySlots = partitions * (n - (maxCount - 1));
    int available = (int)tasks.size() - maxFreq * maxCount;
    int idles = max(0, emptySlots - available);
    return (int)tasks.size() + idles;
}

int main() {
    vector<char> t1 = {'A','A','A','B','B','B'};
    assert(leastInterval(t1, 2) == 8);

    vector<char> t2 = {'A','A','A','B','B','B'};
    assert(leastInterval(t2, 0) == 6);

    vector<char> t3 = {'A','A','A','A','A','A','B','C','D','E','F','G'};
    assert(leastInterval(t3, 2) == 16);

    cout << "All tests passed!" << endl;
    return 0;
}

File: Languages/python/algorithms-dsa/leetcode/greedy/task_scheduler.py
"""
LeetCode #621 - Task Scheduler
Topic: Greedy
Difficulty: Medium

Find minimum intervals needed using frequency counting and idle slot calculation.

Time Complexity: O(n)
Space Complexity: O(1) - at most 26 tasks
"""
from collections import Counter


def least_interval(tasks: list[str], n: int) -> int:
    counts = Counter(tasks)
    max_freq = max(counts.values())
    max_count = sum(1 for v in counts.values() if v == max_freq)
    result = (max_freq - 1) * (n + 1) + max_count
    return max(result, len(tasks))


if __name__ == "__main__":
    assert least_interval(["A", "A", "A", "B", "B", "B"], 2) == 8
    assert least_interval(["A", "A", "A", "B", "B", "B"], 0) == 6
    assert least_interval(["A", "A", "A", "A", "B", "B", "C", "C"], 2) == 10
    print("All tests passed!")
